sympy/core/power.py

```diff
@@ -621,7 +621,7 @@ def _eval_is_imaginary(self):
                 rat = self.exp.is_rational
                 if not rat:
                     return rat
-                if self.exp.is_integer:
+                if self.exp.is_integer and self.base.is_zero is not None:
                     return False
```

Comment:
False here should be `self.base.is_zero`.

Reply:
Will apply the changes ASAP. Thanks.

Reply:
Sir, the testcases are still failing. I applied bin/test core and around 9 testcases failed after I applied the given changes.
@smichr

Reply:
> False here should be `self.base.is_zero`.
here return should be self.base.is_zero instead of return False. Right??

Reply:
Please help to resolve this issue.
@oscarbenjamin @smichr

Reply:
I would rewrite this section. We already know here that the base and the exponent are extended_real. With that in mind when can `x**y` be imaginary? We will have `x**y` imaginary iff x is negative (finite, nonzero) and y is rational m/n with m odd and n 2. Therefore (untested):
```python
if self.base.is_extended_real and self.exp.is_extended_real:
    return fuzzy_and([self.base.is_negative,
                      fuzzy_not(self.exp.is_integer),
                      (2*self.exp).is_odd])
```
We can do that lazily with:
```python
if self.base.is_extended_real and self.exp.is_extended_real:
    def is_imaginary_iff():
        yield self.base.is_negative
        yield fuzzy_not(self.exp.is_integer)
        yield (2*self.exp).is_odd
    return fuzzy_and(is_imaginary_iff())
```

Reply:
I guess the is_integer condition isn't needed if we already know that `(2*exp)` is odd.

Reply:
Sir, the testcases are still failing. I applied bin/test core and around 9 testcases failed after I applied the given changes.
@smichr @oscarbenjamin

Reply:
I can't see what the failures are. Can you resolve the merge conflicts and then push the changes to this PR?

Reply:
> I can't see what the failures are. Can you resolve the merge conflicts and then push the changes to this PR?

So shall I apply the changes mentioned above and push them to this PR?

Reply:
If the changes seem correct to you then yes, push them here.

First you'll need to rebase or merge master to resolve the merge conflicts though.

Reply:
No the changes are incorrect. Just for you to see what are the errors I can push them.

Reply:
You don't have to push them. Just explain *why* they are incorrect.

Reply:
The changes were suggested by @smichr , tough for me to explain why they were not passing the tests, but i will mention which are the testcases that are failing.

Reply:
These are the **testcases which are failing**:

_________ sympy\core\tests\test_arit.py:test_Add_is_negative_positive __________
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_arit.py", line 856, in test_Add_is_negative_positive
    assert z.is_zero
AssertionError
________________________________________________________________________________
________________ sympy\core\tests\test_arit.py:test_Pow_is_real ________________
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_arit.py", line 994, in test_Pow_is_real
    assert (x**2).is_real is True
AssertionError
________________________________________________________________________________
______ sympy\core\tests\test_arit.py:test_Pow_is_nonpositive_nonnegative _______
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_arit.py", line 1208, in test_Pow_is_nonpositive_nonnegative
    assert (x**(4*k)).is_nonnegative is True
AssertionError
________________________________________________________________________________
____________________ sympy\core\tests\test_arit.py:test_Mod ____________________
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_arit.py", line 1600, in test_Mod
    assert Mod(Mod(x, t), 2*t) == Mod(x, t)
AssertionError
________________________________________________________________________________
______________ sympy\core\tests\test_arit.py:test_issue_8247_8354 ______________
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_arit.py", line 2006, in test_issue_8247_8354
    assert z.is_positive is False  # it's 0
AssertionError
________________________________________________________________________________
__________ sympy\core\tests\test_assumptions.py:test_Mul_is_algebraic __________
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_assumptions.py", line 801, in test_Mul_is_algebraic
    assert (a*b).is_algebraic
AssertionError
________________________________________________________________________________
__________ sympy\core\tests\test_assumptions.py:test_Pow_is_algebraic __________
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_assumptions.py", line 821, in test_Pow_is_algebraic
    assert (a**r).is_algebraic
AssertionError
________________________________________________________________________________
________ sympy\core\tests\test_assumptions.py:test_special_is_rational _________
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_assumptions.py", line 947, in test_special_is_rational
    assert (sqrt(3) + S.Pi).is_rational is False
AssertionError
________________________________________________________________________________
______________ sympy\core\tests\test_complex.py:test_issue_11518 _______________
Traceback (most recent call last):
  File "c:\users\sachin\github\sympy\sympy\core\tests\test_complex.py", line 218, in test_issue_11518
    assert conjugate(r) == r
AssertionError

 tests finished: 1720 passed, 9 failed, 65 skipped, 43 expected to fail,
in 39.01 seconds

Reply:
Is that from the change proposed by @smichr or the change I proposed above?

Reply:
The change you proposed, because i rewrote the entire section as you said.

Reply:
I am unable to find any solution. Please review the issue @smichr @oscarbenjamin

Reply:
Working on the assumptions code is difficult especially on the really core classes like `Pow`. Part of the problem is that often when you fix a bug in one place it will show up a different bug somewhere else. That makes it hard to know what the cause of test failures can be.

Reply:
I truly agree with you, so how should we proceed?